lib/Echidna/UI.hs

```diff
@@ -139,15 +148,22 @@ ui :: ( MonadCatch m, MonadRandom m, MonadReader x m, MonadUnliftIO m
    -> [SolTest] -- ^ Tests to evaluate
    -> Maybe GenDict
    -> m Campaign
-ui v w ts d = let xfer e = use hasLens >>= \c -> isDone c >>= ($ e c) . bool id forever in do
-  let d' = fromMaybe defaultDict d
-  s <- (&&) <$> isTerminal <*> view (hasLens . dashboard)
-  g <- view (hasLens . to seed)
-  let g' = fromMaybe (d' ^. defSeed) g
-  c <- if s then do bc <- liftIO $ newBChan 100
-                    t <- forkIO $ campaign (xfer $ liftIO . writeBChan bc) v w ts d >> pure ()
-                    a <- monitor (killThread t)
-                    liftIO (customMain (mkVty defaultConfig) (Just bc) a $ Campaign mempty mempty d' False)
-            else campaign (pure ()) v w ts d
-  liftIO . putStrLn =<< view (hasLens . finished) <*> pure c <*> pure g'
-  return c
+ui v w ts d = let xfer bc = use hasLens >>= liftIO . writeBChan bc
+                  d' = fromMaybe defaultDict d
+                  getSeed = view $ hasLens . to seed . non (d' ^. defSeed) in do
+  bc <- liftIO $ newBChan 100
+  t    <- forkIO (void $ campaign (xfer bc) v w ts d)
+  dash <- liftM2 (&&) isTerminal $ view (hasLens . dashboard)
+  done <- views hasLens $ \cc -> flip runReader (cc :: CampaignConf) . isDone
+  app  <- customMain (mkVty defaultConfig) (Just bc) <$> monitor
+  time <- views (hasLens . maxTime) . maybe (fmap Just) $ timeout . (* 1000000)
+  res  <-  liftIO . time $ if dash
+    then fst <$> app (defaultCampaign, Uninitialized)
+    else let go = readBChan bc >>= \c -> if done c then pure c else go in go
```

Comment:
I'm wondering how much CPU time is consumed by the loop here. Maybe we could put the main thread to sleep while waiting for worker to signal completion/timeout? Here is an example how to wake up a thread with `MVar` https://stackoverflow.com/questions/28169297/how-to-force-main-thread-to-wait-for-all-its-child-threads-finish-in-haskell, however this might require a bigger change.